### PR TITLE
feat(API): remove API endpoints that redirect clients to newer API versioned endpoints

### PR DESCRIPTION
**Pull request by aali309**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] [Signed all commits using a GPG signature](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#gpg-commit-signature-verification)

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes: #583 

## Description of the change:
*This change allows an environment variable to be configured so that...*

## Motivation for the change:
*This change is helpful because users may want to...*

## How to manually test:
1. *Run CRYOSTAT_IMAGE=quay.io... bash smoketest.bash...*
2. *...*

### CODE REVIEW

File: .github/workflows/pr-ci.yaml [hunk 1]
```diff
@@ -276,6 +276,9 @@ jobs:
   auto-commit-schemas:
     needs: [compare-graphql-schema, compare-openapi-schema, checkout-branch, update-schemas]
     runs-on: ubuntu-latest
+    permissions:
+      contents: write
+      pull-requests: write
```
Comment (andrewazores) [inline .github/workflows/pr-ci.yaml:281]: Is this needed? Should it be a separate PR?

File: src/main/java/io/cryostat/recordings/Recordings.java [hunk 2]
```diff
@@ -541,7 +508,13 @@ public Uni<Response> createSnapshotV1(@RestPath URI connectUrl) throws Exception
...
                 .transform(
                         recording ->
-                                Response.status(Response.Status.OK).entity(recording.name).build())
+                                Response.status(Response.Status.OK)
+                                        .entity(
+                                                new JsonObject()
+                                                        .put("name", recording.name)
+                                                        .put("remoteId", recording.remoteId)
+                                                        .encode())
+                                        .build())
```
Comment (andrewazores) [inline src/main/java/io/cryostat/recordings/Recordings.java:517]: This changes breaks the V1 API.

File: src/main/java/io/cryostat/recordings/Recordings.java [hunk 3]
```diff
@@ -558,10 +531,7 @@ public Uni<Response> createSnapshotV2(@RestPath URI connectUrl) throws Exception
                 .transform(
                         recording ->
                                 Response.status(Response.Status.CREATED)
-                                        .entity(
-                                                V2Response.json(
-                                                        Response.Status.CREATED,
-                                                        recordingHelper.toExternalForm(recording)))
+                                        .entity(recordingHelper.toExternalForm(recording))
```
Comment (andrewazores) [inline src/main/java/io/cryostat/recordings/Recordings.java:534]: We want to remove the `V2Response` format, but doing this should also mean removing or renaming any endpoints that return it. So this `/api/v2` endpoint should either be removed, or it should be renamed to `/api/v3` if it will return this simplified JSON response format directly.

File: src/main/java/io/cryostat/recordings/Recordings.java [hunk 4]
```diff
@@ -598,7 +568,8 @@ public Response createRecording(
             @RestForm String recordingName,
             @RestForm String events,
             @RestForm Optional<String> replace,
-            // restart param is deprecated, only 'replace' should be used and takes priority if both
+            // restart param is deprecated, only 'replace' should be used and takes priority
+            // if both
```
Comment (andrewazores) [inline src/main/java/io/cryostat/recordings/Recordings.java:572]: Funny linebreaks here again.

**Review by andrewazores — COMMENTED** [review]
Looking mostly good. Before this is merged I'd also like to see corresponding PRs for -web and -agent to make sure they are up to speed with the API changes, if any are necessary.

Maybe it would be best to create an upstream feature branch for the overall API v4 cleanup work?

File: .github/workflows/pr-ci.yaml (hunk 1 above)
Comment (aali309) [reply .github/workflows/pr-ci.yaml:281]: Yes, I will open a separate PR for this.

File: src/main/java/io/cryostat/credentials/Credentials.java [hunk 5]
```diff
@@ -41,38 +42,58 @@
...
+                                            logger.warn(e);
+                                            return null;
+                                        }
+                                    })
+                            .filter(Objects::nonNull)
+                            .toList();
+
+            JsonObject response = new JsonObject();
+            response.put(
+                    "meta", new JsonObject().put("type", "application/json").put("status", "OK"));
```
Comment (andrewazores) [inline src/main/java/io/cryostat/credentials/Credentials.java:72]: This is the same thing as as V2Response, just manually encoded instead of using a utility structure. I think this endpoint should just `return Response.ok(results).type(MediaType.APPLICATION_JSON).build();` for the new API format.

File: src/main/java/io/cryostat/credentials/Credentials.java [hunk 6]
```diff
@@ -41,38 +42,58 @@
...
+        try {
+            Credential credential = Credential.find("id", id).singleResult();
+            Map<String, Object> result = safeMatchedResult(credential, targetMatcher);
+            return Response.ok(result).type(MediaType.APPLICATION_JSON).build();
+        } catch (ScriptException e) {
+            logger.error("Error retrieving credential", e);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        } catch (Exception e) {
+            logger.error("Credential not found", e);
+            return Response.status(Response.Status.NOT_FOUND).build();
```
Comment (andrewazores) [inline src/main/java/io/cryostat/credentials/Credentials.java:95]: I think the NOT_FOUND case should already be handled here:

https://github.com/cryostatio/cryostat/blob/e6457be1994acd310cbbc03734a0c577b5ae817c/src/main/java/io/cryostat/ExceptionMappers.java#L45

Comment (andrewazores) [reply src/main/java/io/cryostat/credentials/Credentials.java:95]: And if for some reason not, then this final catch-clause should catch something more specific than `Exception`, since there could be a lot of other kinds of unexpected RuntimeException that may occur and which should not map to a NOT_FOUND status.

File: src/main/java/io/cryostat/discovery/CustomDiscovery.java [hunk 7]
```diff
@@ -110,16 +108,16 @@ public Response create(
         } catch (Exception e) {
             logger.error("Target validation failed", e);
             return Response.status(Response.Status.BAD_REQUEST)
-                    .entity(V2Response.json(Response.Status.BAD_REQUEST, e))
+                    .entity(Map.of(Response.Status.BAD_REQUEST, e.getMessage()))
```
Comment (andrewazores) [inline src/main/java/io/cryostat/discovery/CustomDiscovery.java:111]: I think just `e.getMessage()` would suffice as the response body here.

File: src/main/java/io/cryostat/discovery/CustomDiscovery.java [hunk 8]
```diff
@@ -156,7 +154,7 @@ Response doV2Create(
             if (!uriUtil.validateUri(target.connectUrl)) {
                 return Response.status(Response.Status.BAD_REQUEST)
                         .entity(
-                                String.format(
+                                Map.of(
```
Comment (andrewazores) [inline src/main/java/io/cryostat/discovery/CustomDiscovery.java:157]: This doesn't seem right

File: src/main/java/io/cryostat/discovery/CustomDiscovery.java [hunk 9]
```diff
@@ -165,9 +163,7 @@ Response doV2Create(
 
             if (Target.find("connectUrl", target.connectUrl).singleResultOptional().isPresent()) {
                 return Response.status(Response.Status.BAD_REQUEST.getStatusCode())
-                        .entity(
-                                V2Response.json(
-                                        Response.Status.BAD_REQUEST, "Duplicate connection URL"))
+                        .entity(Map.of(Response.Status.BAD_REQUEST, "Duplicate connection URL"))
```
Comment (andrewazores) [inline src/main/java/io/cryostat/discovery/CustomDiscovery.java:166]: Similar to above, this can just be the `"Duplication connection URL"` string, no need to create a Map.

File: src/main/java/io/cryostat/expressions/MatchExpressions.java [hunk 10]
```diff
@@ -63,7 +65,34 @@ public V2Response test(RequestData requestData) throws ScriptException {
...
+        matched.targets()
+                .forEach(
+                        target -> {
+                            JsonObject targetJson = new JsonObject();
+                            targetJson.put("connectUrl", target.connectUrl);
+                            targetJson.put("alias", target.alias);
+                            targetsJson.add(targetJson);
+                        });
+
+        return new JsonObject()
```
Comment (andrewazores) [inline src/main/java/io/cryostat/expressions/MatchExpressions.java:85]: Here again, this is just manually creating the equivalent to a V2Response, which is mainly what we're trying to avoid doing anymore. I think the response body here would just be the `MatchedExpression` object.

File: src/main/java/io/cryostat/JsonRequestFilter.java [hunk 11]
```diff
@@ -39,11 +39,11 @@ public class JsonRequestFilter implements ContainerRequestFilter {
...
     static final Set<String> allowedPathPatterns =
             Set.of(
-                    "/api/v2.2/discovery",
-                    "/api/v2/rules/[\\w]+",
+                    "/api/v4/discovery",
+                    "/api/v4/rules/[\\w]+",
                     "/api/beta/matchExpressions",
-                    "/api/v2.2/graphql",
-                    "/api/v3/graphql");
+                    "/api/v2/graphql",
```
Comment (andrewazores) [inline src/main/java/io/cryostat/JsonRequestFilter.java:45]: `v2/graphql` can be removed now, right?

Comment (aali309) [reply src/main/java/io/cryostat/JsonRequestFilter.java:45]: yes

Comment (aali309) [reply src/main/java/io/cryostat/JsonRequestFilter.java:45]: v1 path should also be v4? @andrewazores

Comment (andrewazores) [reply src/main/java/io/cryostat/JsonRequestFilter.java:45]: Yes please, everything should become v4 in this release.

File: src/main/java/io/cryostat/credentials/CredentialCheck.java [hunk 12]
```diff
@@ -42,7 +40,7 @@ public class CredentialCheck {
     @Blocking
     @RolesAllowed("read")
     @Path("/api/beta/credentials/{connectUrl}")
-    public Uni<V2Response> checkCredentialForTarget(
+    public Uni<CredentialTestResult> checkCredentialForTarget(
```
Comment (andrewazores) [inline src/main/java/io/cryostat/credentials/CredentialCheck.java:43]: @aali309 take a look at this change please - in this case it's better to just return the result type directly, rather than wrapping it in a JAX-RS Response. Quarkus will automatically set the Content-Type response header anyway, so this saves a couple of lines of code but also makes sure that the content negotiation between the Cryostat server and the user's browser properly determines the response format and header value.

Comment (andrewazores) [reply src/main/java/io/cryostat/credentials/CredentialCheck.java:43]: This same simplification should be applied elsewhere too, unless there is a particular reason to wrap the result in a Response, but I think most of the time there won't be.

File: src/main/java/io/cryostat/credentials/Credentials.java [hunk 13]
```diff
@@ -89,7 +111,9 @@ public RestResponse<Void> create(
         credential.username = username;
         credential.password = password;
         credential.persist();
-        return ResponseBuilder.<Void>created(URI.create("/api/v2.2/credentials/" + credential.id))
+        return ResponseBuilder.<Credential>created(
+                        uriInfo.getAbsolutePathBuilder().path(Long.toString(credential.id)).build())
```
Comment (andrewazores) [inline src/main/java/io/cryostat/credentials/Credentials.java:115]: @aali309 see this as well - I found this nice way to set the Location header in responses relative to the current request path. This pattern will apply most of the time to `@POST` endpoints, and it saves us having to update so many hardcoded strings in the future.

File: src/main/java/io/cryostat/discovery/CustomDiscovery.java [hunk 14]
```diff
@@ -143,10 +139,10 @@ public Response create(
             credential.password = password;
         }
 
-        return doV2Create(target, Optional.ofNullable(credential), dryrun, storeCredentials);
+        return doV4Create(target, Optional.ofNullable(credential), dryrun, storeCredentials);
     }
 
-    Response doV2Create(
+    Response doV4Create(
```
Comment (andrewazores) [inline src/main/java/io/cryostat/discovery/CustomDiscovery.java:145]: I think this can just be renamed `doCreate` now.

File: src/main/java/io/cryostat/discovery/CustomDiscovery.java [hunk 15]
```diff
@@ -165,9 +161,7 @@ Response doV2Create(
 
             if (Target.find("connectUrl", target.connectUrl).singleResultOptional().isPresent()) {
                 return Response.status(Response.Status.BAD_REQUEST.getStatusCode())
-                        .entity(
-                                V2Response.json(
-                                        Response.Status.BAD_REQUEST, "Duplicate connection URL"))
+                        .entity("Duplicate connection URL")
```
Comment (andrewazores) [inline src/main/java/io/cryostat/discovery/CustomDiscovery.java:164]: I think this `Response.status()` call can become `throw new BadRequestException("Duplicate connection URL")`, and similar for the next few lower down as well.

**Comment by andrewazores** [comment]
After merging main into `api-v4` and updating this branch against that, we have the supposed conflicts above. I think it's mostly just GitHub recognizing the history incorrectly. If it still persists after merging this into `api-v4` and opening a PR from that back into `main` then I'll deal with it then.

In the meantime, as I'm working through this and adding tests, https://github.com/cryostatio/cryostat/pull/614 would be nice to have - it makes it possible to reset the database between tests or after test classes, so it's much easier to ensure that things are isolated and starting from a clean slate without having to do a lot of `try-catch-finally` and repeated cleanup hooks etc.

File: .github/workflows/pr-ci.yaml [hunk 16]
```diff
@@ -276,9 +276,6 @@ jobs:
   auto-commit-schemas:
     needs: [compare-graphql-schema, compare-openapi-schema, checkout-branch, update-schemas]
     runs-on: ubuntu-latest
-    permissions:
-      contents: write
-      pull-requests: write
```
Comment (mwangggg) [inline .github/workflows/pr-ci.yaml:281]: The permissions shouldn't be removed - I'm assuming this is because of Atif's new PR to add them in.

**Comment by Josh-Matsuoka** [comment]
First readthrough of the changes looks good, I'll test this with the web client and agent changes and make sure everything works as expected, are there any additional PRs connected to this one that should be tested at the same time other than the ones connected to https://github.com/cryostatio/cryostat/issues/550?

**Comment by andrewazores** [comment]
I think it's just the three: this one, the web one, and [the agent one](https://github.com/cryostatio/cryostat-agent/pull/469).

**Review by Josh-Matsuoka — APPROVED** [review]
